src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer.cs

```diff
@@ -750,34 +821,59 @@ private DiagnosticData UpdatePosition(DiagnosticData diagnostic, SyntaxTree tree
                 helpLink: diagnostic.HelpLink);
         }
 
-        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
+        private static IEnumerable<DiagnosticData> GetDiagnosticData(Document document, SyntaxTree tree, TextSpan? span, IEnumerable<Diagnostic> diagnostics)
         {
-            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, span)).Select(d => DiagnosticData.Create(document, d)) : null;
+            return diagnostics != null ? diagnostics.Where(dx => ShouldIncludeDiagnostic(dx, tree, span)).Select(d => DiagnosticData.Create(document, d)) : null;
         }
 
-        private static bool ShouldIncludeDiagnostic(Diagnostic diagnostic, TextSpan? span)
+        private static bool ShouldIncludeDiagnostic(Diagnostic diagnostic, SyntaxTree tree, TextSpan? span)
         {
             if (diagnostic == null)
             {
                 return false;
             }
 
-            if (span == null)
+            if (diagnostic.Location == null || diagnostic.Location == Location.None)
             {
-                return true;
+                return false;
             }
 
-            if (diagnostic.Location == null)
+            if (diagnostic.Location.SourceTree != tree)
             {
                 return false;
             }
 
+            if (span == null)
+            {
+                return true;
+            }
+
             return span.Value.Contains(diagnostic.Location.SourceSpan);
         }
 
         private static IEnumerable<DiagnosticData> GetDiagnosticData(Project project, IEnumerable<Diagnostic> diagnostics)
         {
-            return diagnostics != null ? diagnostics.Select(d => DiagnosticData.Create(project, d)) : null;
+            if (diagnostics == null)
+            {
+                yield break;
+            }
+
+            foreach (var diagnostic in diagnostics)
+            {
+                if (diagnostic.Location == null || diagnostic.Location == Location.None)
```

Comment:
@mavasani just realized that we can't assume any of properties of diagnostic not returning null values... we probably need to go through where we access them and make sure we don't always assume it is something we created.